src/transformers.ts

```diff
@@ -0,0 +1,140 @@
+      );
+    }
+    getCustomTransformers = customerTransformers;
+  }
+  return getCustomTransformers(program);
+}
+
+/**
+ * @returns a ts.CustomTransformers merged from tr1 and tr2 (does not check for duplicates)
+ */
+function mergeTransformers(tr1: ts.CustomTransformers | undefined, tr2: ts.CustomTransformers | undefined): ts.CustomTransformers | undefined {
+  if (tr1) {
+    const result: ts.CustomTransformers = {
+      after: [],
+      before: [],
+      afterDeclarations: []
+    };
+    //add tr1 values in result
+    if (tr1.after) {
+      result.after!.push(...tr1.after);
+    }
+    if (tr1.before) {
+      result.before!.push(...tr1.before);
+    }
+    if (tr1.afterDeclarations) {
+      result.afterDeclarations!.push(...tr1.afterDeclarations);
+    }
+
+    //add tr2 values
+    if (tr2?.after) {
+      result.after!.push(...tr2.after);
+    }
+    if (tr2?.before) {
+      result.before!.push(...tr2.before);
+    }
+    if (tr2?.afterDeclarations) {
+      result.afterDeclarations!.push(...tr2.afterDeclarations);
+    }
+
+    return result;
+  } else {
+    return tr2;
+  }
+}
+
+/**
+ * Patches typescript program to pass custom transformers to the program.emit function.
+ * In 2021, this is the only way since it is not possible to pass transformers to SolutionBuilder
+ * SolutionBuilder is used when watch is enabled
+ */
+let patched = false;
+export function installTransformers(loaderOptions: LoaderOptions) {
+  if (loaderOptions.getCustomTransformers === undefined) {
+    return; //no need to patch if there is no transformer
+  }
+  if (patched === true) {
+    return; //do not patch twice
+  }
+  patched = true;
+  const originalCreateProgram = ts.createProgram;
+
+  //override createProgram in order to override emit function and use overrideTransformers
+  function createProgram(
+    rootNamesOrOptions: ReadonlyArray<string> | ts.CreateProgramOptions,
+    options?: ts.CompilerOptions,
+    host?: ts.CompilerHost,
+    oldProgram?: ts.Program,
+    configFileParsingDiagnostics?: ReadonlyArray<ts.Diagnostic>
+  ): ts.Program {
+    let program: ts.Program;
+    if (Array.isArray(rootNamesOrOptions)) {
+      program = originalCreateProgram(
+        rootNamesOrOptions as ReadonlyArray<string>,
+        options!,
+        host,
+        oldProgram,
+        configFileParsingDiagnostics
+      );
+    } else {
+      program = originalCreateProgram(
+        rootNamesOrOptions as ts.CreateProgramOptions
+      );
+    }
+
+    const transformers: ts.CustomTransformers | undefined = getCustomTransformers(loaderOptions, program);
+
+    const originalEmit = program.emit;
+
+    function newEmit(
+      targetSourceFile?: ts.SourceFile,
+      writeFile?: ts.WriteFileCallback,
+      cancellationToken?: ts.CancellationToken,
+      emitOnlyDtsFiles?: boolean,
+      customTransformers?: ts.CustomTransformers
+    ): ts.EmitResult {
+      /* Invoke TS emit */
+      const result: ts.EmitResult = originalEmit(
```

Comment:
This will have some bad effects .. emit is used for emitting dts files to determine shape of the file as well and take internal parameters and that signature is internal to typescript .. also I don’t think you want transformers for that reason .. 

I have been thinking about your issue and was wondering if build could take getTransformers as function that returns transformers based on project path.. haven’t had chance to investigate that part yet but this approach is definitely not correct

Reply:
Oh thanks for that @sheetalkamat - very helpful!

Reply:
Yes, if we can pass customTransformers to build in SolutionBuilder, this solves the problem.
the done function on invalidatedProject already has the customTransformers parameter.
![image](https://user-images.githubusercontent.com/6431249/117301061-870dbe00-ae7a-11eb-8f54-e5c0699c2fb3.png)

A function by projet path is ok, anyway in ts-loader there is only one global configuration so we would always pass the same transformers. A good solution would be to do like ttypescript (https://github.com/cevek/ttypescript) and read the configuration from tsconfig.json

Reply:
https://github.com/microsoft/TypeScript/pull/43984 add the typescript side of the API